support-services/src/main/scala/com/gu/support/promotions/PromotionService.scala

```diff
@@ -26,10 +26,12 @@ class PromotionService(config: PromotionsConfig, maybeCollection: Option[Promoti
     }
 
   def findPromotions(promoCodes: List[PromoCode]): List[PromotionWithCode] =
-    allWith6For6
-      .foldLeft(List.empty[PromotionWithCode]) { (acc, promotion) =>
-        val maybeCode = promoCodes.intersect(promotion.promoCodes.toList).headOption
-        maybeCode.map(code => acc :+ PromotionWithCode(code, promotion)).getOrElse(acc)
+    promoCodes
+      .foldLeft(List.empty[PromotionWithCode]) { (acc, promoCode) =>
+        allWith6For6.find(promo => promo.promoCodes.exists(_ == promoCode)) match {
+          case Some(promotion) => acc :+ PromotionWithCode(promoCode, promotion)
+          case None => acc
+        }
```

Comment:
so before the time complxity was
O(len(allWith6For6) * intersect)
now it would be
O(len(promoCodes) * len(allWith6For6))
so something like n log n to n^2

As long as we are happy the `promoCodes` list is bounded in some way that's probably fine, if it's going to build up over time it is lless so.

Reply:
even if duplicated need to be retained we could do something like this (not compiled)
```suggestion
// could be precomputed into a static/const
    val promosPerCode: Map[PromoCode, List[Promotion]] = allWith6For6.flatMap(promo.promoCodes.map((_, promo))).groupBy(_._1).mapValues(_.map(pC => PromotionWithCode(pC._1, pc._2)).toMap

    promoCodes.flatMap(promosPerCode.get)

```